File: backend/schemas/agents.py
"""Agent-related schemas."""

from datetime import datetime
from typing import Optional

from i18n.serializers import serialize_bool as _serialize_bool
from i18n.serializers import serialize_utc_datetime as _serialize_utc_datetime
from pydantic import BaseModel, field_serializer


class AgentBase(BaseModel):
    name: str
    group: Optional[str] = None
    config_file: Optional[str] = None
    profile_pic: Optional[str] = None
    in_a_nutshell: Optional[str] = None
    characteristics: Optional[str] = None
    recent_events: Optional[str] = None
    interrupt_every_turn: bool = False
    priority: int = 0


class AgentCreate(AgentBase):
    """
    Create an agent with either:
    1. config_file: Load in_a_nutshell/characteristics/recent_events from file
    2. in_a_nutshell/characteristics/recent_events: Provide directly
    The system_prompt will be built automatically.
    """

    pass


class AgentUpdate(BaseModel):
    """Update agent's runtime fields: nutshell, characteristics, or recent events."""

    profile_pic: Optional[str] = None
    in_a_nutshell: Optional[str] = None
    characteristics: Optional[str] = None
    recent_events: Optional[str] = None


class Agent(AgentBase):
    id: int
    system_prompt: str  # The built system prompt
    session_id: Optional[str] = None
    created_at: datetime

    @field_serializer("created_at")
    def serialize_created_at(self, dt: datetime, _info):
        return _serialize_utc_datetime(dt)

    @field_serializer("interrupt_every_turn")
    def serialize_interrupt_every_turn(self, value: int, _info):
        return _serialize_bool(value)

    class Config:
        from_attributes = True


__all__ = [
    "AgentBase",
    "AgentCreate",
    "AgentUpdate",
    "Agent",
]

File: backend/src/schemas/agents.ts
/** Agent request/response schemas. */

import { z } from 'zod'
import type { Agent as AgentRow } from '../db/schema'
import {
  isoDatetime,
  optionalString,
  pydanticBool,
  pydanticInt,
  requiredTimestamp,
  serializeBool,
} from './common'

// `interrupt_every_turn` and `priority` are behaviour settings a
// `group_config.yaml` can override, not per-agent data.
export const AgentBase = z.object({
  name: z.string(),
  group: optionalString(),
  config_file: optionalString(),
  profile_pic: optionalString(),
  in_a_nutshell: optionalString(),
  characteristics: optionalString(),
  recent_events: optionalString(),
  interrupt_every_turn: pydanticBool().default(false),
  priority: pydanticInt().default(0),
})

export type AgentBase = z.infer<typeof AgentBase>

// From a `config_file` or the three markdown fields inline. `system_prompt` is
// built server-side, so it is on the response and not here.
export const AgentCreate = AgentBase

export type AgentCreate = z.infer<typeof AgentCreate>

/** The runtime-editable subset: the picture and the three markdown files. */
export const AgentUpdate = z.object({
  profile_pic: optionalString(),
  in_a_nutshell: optionalString(),
  characteristics: optionalString(),
  recent_events: optionalString(),
})

export type AgentUpdate = z.infer<typeof AgentUpdate>

/** The agent response: base fields first, then the four declared here. */
export const Agent = AgentBase.extend({
  id: pydanticInt(),
  /** Built from the markdown files by the prompt builder, never client-supplied. */
  system_prompt: z.string(),
  session_id: optionalString(),
  created_at: isoDatetime(),
})

export type Agent = z.infer<typeof Agent>

/**
 * Drizzle row → `Agent` response. `session_id` is always `null` — there is no
 * such column, SDK session ids live in `room_agent_sessions` — but the wire
 * contract carries it. `world_name` is deliberately not exposed.
 */
export function toAgent(row: AgentRow): Agent {
  return {
    name: row.name,
    group: row.group,
    config_file: row.configFile,
    profile_pic: row.profilePic,
    in_a_nutshell: row.inANutshell,
    characteristics: row.characteristics,
    recent_events: row.recentEvents,
    interrupt_every_turn: serializeBool(row.interruptEveryTurn),
    priority: row.priority ?? 0,
    id: row.id,
    system_prompt: row.systemPrompt,
    session_id: null,
    created_at: requiredTimestamp(row.createdAt, 'Agent', 'created_at'),
  }
}
